crates/ra_hir/src/ty.rs

```diff
@@ -1042,12 +1046,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             },
             Expr::Tuple { exprs } => {
                 let mut ty_vec = Vec::with_capacity(exprs.len());
-                for arg in exprs.iter() {
-                    ty_vec.push(self.infer_expr(*arg, &Expectation::none())?);
+                for expr in exprs.iter() {
+                    ty_vec.push(self.infer_expr(*expr, &Expectation::none())?);
                 }
 
                 Ty::Tuple(Arc::from(ty_vec))
             }
+            Expr::Array { exprs } => {
+                let mut elem_ty = match &expected.ty {
+                    Ty::Slice(inner) | Ty::Array(inner) => Ty::clone(&inner),
+                    _ => self.new_type_var(),
+                };
+
+                for expr in exprs.iter() {
+                    elem_ty = self.infer_expr(*expr, &Expectation::has_type(elem_ty.clone()))?;
```

Comment:
I don't think you have to assign elem_ty here. You should be able to keep using the elem_ty above:
```suggestion
                    self.infer_expr(*expr, &Expectation::has_type(elem_ty.clone()))?;
```

(and remove the `mut`).